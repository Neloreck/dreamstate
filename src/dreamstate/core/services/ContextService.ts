import { queryData } from "@/dreamstate/core/queries/queryData";
import { emitSignal } from "@/dreamstate/core/signals/emitSignal";
import {
  IBaseSignal,
  IOptionalQueryRequest,
  TAnyObject,
  TDreamstateService,
  TQueryType,
  TSignalType
} from "@/dreamstate/types";

export abstract class ContextService<S extends TAnyObject = TAnyObject> {

  /**
   * Should core destroy store instance after observers removal or preserve it for application lifespan.
   * Singleton objects will never be destroyed once created.
   * Non-singleton objects are destroyed if all observers are removed.
   */
  protected static IS_SINGLE: boolean = false;

  public constructor(initialState?: S) {
  }

  /**
   * Lifecycle.
   * First provider was injected into DOM.
   */
  protected onProvisionStarted(): void {}

  /**
   * Lifecycle.
   * Last provider was removed from DOM.
   */
  protected onProvisionEnded(): void {}

  /**
   * Emit signal for other managers and subscribers.
   */
  protected emitSignal<T extends TSignalType = TSignalType, D = undefined>(
    baseSignal: IBaseSignal<T, D>
  ): Promise<void> {
    return emitSignal(baseSignal, this.constructor as TDreamstateService<S>);
  }

  /**
   * Send context query to retrieve data from @OnQuery method with required types.
   */
  protected queryData<
    R extends any,
    D extends any,
    T extends TQueryType,
    Q extends IOptionalQueryRequest<D, T> | Array<IOptionalQueryRequest>
    >(
    queryRequest: Q
  ) {
    return queryData<R, D, T, Q>(queryRequest);
  }

  /**
   * Send sync context query to retrieve data from @OnQuery method with required types.
   */
  protected queryDataSync<
    R extends any,
    D extends any,
    T extends TQueryType,
    Q extends IOptionalQueryRequest<D, T>
    >(
    queryRequest: Q
  ) {
    return queryData<R, D, T, Q>(queryRequest);
  }

}
