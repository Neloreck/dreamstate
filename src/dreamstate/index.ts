/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export {
  ILoadable as Loadable,
  TNested as Nested,
  TComputed as Computed,
  TStateSetter as StateSetter,
  TConsumable as Consumable,
  ISignalEvent as SignalEvent,
  ISignal as Signal,
  TSignalType as SignalType,
  TQueryResponse as QueryResponse,
  TOptionalQueryResponse as OptionalQueryResponse,
  TQueryRequest as QueryRequest,
  TQueryType as QueryType
} from "@/dreamstate/types";

export { ContextService } from "@/dreamstate/core/services/ContextService";
export { ContextManager } from "@/dreamstate/core/services/ContextManager";

export { getCurrentContext } from "@/dreamstate/core/registry/getCurrentContext";
export { getCurrent } from "@/dreamstate/core/registry/getCurrent";

export { createProvider } from "@/dreamstate/core/provision/createProvider";
export { Provide } from "@/dreamstate/core/provision/Provide";
export { withProvision } from "@/dreamstate/core/provision/withProvision";

export { OnSignal } from "@/dreamstate/core/signals/OnSignal";
export { useSignals } from "@/dreamstate/core/signals/useSignals";
export { unsubscribeFromSignals } from "@/dreamstate/core/signals/unsubscribeFromSignals";
export { subscribeToSignals } from "@/dreamstate/core/signals/subscribeToSignals";
export { emitSignal } from "@/dreamstate/core/signals/emitSignal";

export { OnQuery } from "@/dreamstate/core/queries/OnQuery";
export { queryData } from "@/dreamstate/core/queries/queryData";

export { useManager } from "@/dreamstate/core/consumption/useManager";
export { Consume } from "@/dreamstate/core/consumption/Consume";
export { withConsumption } from "@/dreamstate/core/consumption/withConsumption";

export { Bind } from "@/dreamstate/utils/Bind";
export { createLoadable } from "@/dreamstate/utils/createLoadable";
export { createNested } from "@/dreamstate/utils/createNested";
export { createComputed } from "@/dreamstate/utils/createComputed";
export { createSetter } from "@/dreamstate/utils/createSetter";
