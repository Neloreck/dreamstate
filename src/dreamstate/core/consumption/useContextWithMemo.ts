import { MutableRefObject, useCallback, useEffect, useRef, useState } from "react";

import { CONTEXT_STATES_REGISTRY, EMPTY_ARR } from "@/dreamstate/core/internals";
import { subscribeToManager } from "@/dreamstate/core/registry/subscribeToManager";
import { unsubscribeFromManager } from "@/dreamstate/core/registry/unSubscribeFromManager";
import { IContextManagerConstructor, TUpdateSubscriber } from "@/dreamstate/types";

/**
 * Use manager hook with subscribed updates.
 * Same like common useContext hook, but has memo checks.
 */
export function useContextWithMemo<T extends object, D extends IContextManagerConstructor<T>>(
  Manager: D,
  depsSelector: (context: T) => Array<any>
): T {
  const [ state, setState ] = useState(function() {
    return CONTEXT_STATES_REGISTRY.get(Manager) as T;
  });
  const observed: MutableRefObject<Array<any>> = useRef(depsSelector(state));

  const updateMemoState: TUpdateSubscriber<T> = useCallback(function(nextContext: T): void {
    // Calculate changes like react core does and fire change only after update.
    const nextObserved = depsSelector(nextContext);

    for (let it = 0; it < nextObserved.length; it ++) {
      if (observed.current[it] !== nextObserved[it]) {
        observed.current = nextObserved;
        setState(nextContext);

        return;
      }
    }
  }, EMPTY_ARR);

  useEffect(function() {
    subscribeToManager(Manager, updateMemoState);

    return function() {
      unsubscribeFromManager(Manager, updateMemoState);
    };
  });

  return state;
}
