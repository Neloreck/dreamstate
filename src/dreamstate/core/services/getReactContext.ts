import { Context, createContext } from "react";

import { CONTEXT_REACT_CONTEXTS_REGISTRY } from "@/dreamstate/core/internals";
import { IContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get react context reference for provided ManagerClass.
 * Lazy internal creation only after first access attempt.
 *
 * @param ManagerClass - context manager constructor reference.
 * @param defaultContext - default context value to be applied if manager is not provided.
 * @returns {Context} react context instance with pre-defined default value.
 */
export function getReactContext<S, M extends IContextManagerConstructor<S>>(
  ManagerClass: M,
  defaultContext: Partial<S> | null = null
): Context<S> {
  if (CONTEXT_REACT_CONTEXTS_REGISTRY.has(ManagerClass)) {
    return CONTEXT_REACT_CONTEXTS_REGISTRY.get(ManagerClass) as Context<S>;
  } else {
    const reactContext: Context<S> = createContext(defaultContext as S);

    /**
     * Later providers and consumers in tree will be displayed as
     * 'DS.Class.Provider' or 'DS.Class.Consumer'.
     */
    reactContext.displayName = "DS." + ManagerClass.name;

    CONTEXT_REACT_CONTEXTS_REGISTRY.set(ManagerClass, reactContext);

    return reactContext;
  }
}
