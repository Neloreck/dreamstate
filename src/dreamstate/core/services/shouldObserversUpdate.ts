import { shallowEqualObjects } from "shallow-equal";

import { ActionsStore } from "@/dreamstate/core/storing/ActionsStore";
import { ComputedValue } from "@/dreamstate/core/storing/ComputedValue";
import { NestedStore } from "@/dreamstate/core/storing/NestedStore";
import { TAnyObject } from "@/dreamstate/types";
import { isObject } from "@/dreamstate/utils/typechecking";

/**
 * Compare context manager state diff with shallow check + nested objects check.
 * Respect all meta-fields like NestedStore/ComputedValue/ActionsStore etc.
 *
 * @param previousContext - previous context to match.
 * @param nextContext - next context to match.
 * @returns {boolean} result of context objects comparison, whether observers should update or not.
 */
export function shouldObserversUpdate<T extends TAnyObject>(previousContext: T, nextContext: T): boolean {
  if (!isObject(nextContext)) {
    throw new TypeError(`Context should be non-nullable object, supplied '${typeof nextContext}' type instead.`);
  }

  if (!previousContext) {
    return true;
  }

  return Object.keys(nextContext).some(function(key: string): boolean {
    /**
     * Ignore computed values.
     * Ignore action values.
     * Since nested computed stores are not representing data itself, we should not verify anything there.
     */
    if (nextContext[key] instanceof ComputedValue || nextContext[key] instanceof ActionsStore) {
      return false;
    }

    /**
     * Shallow check for mutable objects created by library.
     * Optimization for sub-states to prevent pollution of context and improve performance.
     * We cannot guess about each object because it is (1) not obvious, (2) can be unwanted and (3) will not work for
     * some objects like native MediaStream/MediaStreamTrack.
     */
    return nextContext[key] instanceof NestedStore
      ? !shallowEqualObjects(nextContext[key], previousContext[key])
      : nextContext[key] !== previousContext[key];
  });
}
