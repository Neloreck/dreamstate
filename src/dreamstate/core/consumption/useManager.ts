import { useContext } from "react";

import { useContextWithMemo } from "@/dreamstate/core/consumption/useContextWithMemo";
import { IContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Use manager hook, higher order wrapper for useContext.
 */
export function useManager<
  T extends TAnyObject,
  D extends IContextManagerConstructor<T, any>
>(
  Manager: D,
  depsSelector?: (context: D["prototype"]["context"]) => Array<any>
): D["prototype"]["context"] {
  if (depsSelector) {
    return useContextWithMemo(Manager, depsSelector);
  } else {
    return useContext(Manager.REACT_CONTEXT);
  }
}
