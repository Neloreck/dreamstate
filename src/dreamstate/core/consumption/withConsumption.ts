import { Consume } from "@/dreamstate/core/consumption/Consume";
import { IConsume } from "@/dreamstate/types";

/**
 * HOC alias for @Consume.
 */
export const withConsumption: IConsume = Consume as IConsume;
