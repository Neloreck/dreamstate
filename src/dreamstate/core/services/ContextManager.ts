import { Context } from "react";

import {
  QUERY_METADATA_SYMBOL,
  SCOPE_SYMBOL,
  SIGNAL_METADATA_SYMBOL,
  SIGNALING_HANDLER_SYMBOL
} from "@/dreamstate/core/internals";
import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { getReactContext } from "@/dreamstate/core/services/getReactContext";
import { shouldObserversUpdate } from "@/dreamstate/core/services/shouldObserversUpdate";
import { processComputed } from "@/dreamstate/core/storing/processComputed";
import { throwOutOfScope } from "@/dreamstate/core/utils/throwOutOfScope";
import {
  IBaseSignal,
  IOptionalQueryRequest,
  ISignalEvent,
  TAnyContextManagerConstructor,
  TAnyObject,
  TConstructorKey,
  TEmptyObject,
  TPartialTransformer,
  TQueryResponse,
  TQuerySubscriptionMetadata,
  TQueryType,
  TSignalListener,
  TSignalSubscriptionMetadata,
  TSignalType
} from "@/dreamstate/types";
import { isFunction } from "@/dreamstate/utils/typechecking";

/**
 * Abstract context manager class.
 * Wraps data and logic part separately from react tree and allows to create global/local storages
 *   that have lifecycle and can be cleaned up and ejected when needed.
 *
 * To provide specific ContextManager classes in react tree check 'createProvider' method.
 * To consume specific ContextManager data in react tree check 'useManager' method.
 *
 * Every class instance is created automatically by dreamstate scope if provision is needed.
 * Every class instance can emit signals/query data in scope where it was created.
 * Every class instance can register method as scope signal listener.
 * Every class instance can register method as scope query data provider.
 * Every class instance is responsible only for specific data part like reducer in redux.
 *   Examples: AuthManager, GraphicsManager, ChatManager, LocalMediaManager etc.
 *
 * Async methods called after manager class unregistering will cause dev warnings and will not affect actual scope.
 */
export abstract class ContextManager<T extends TAnyObject = TEmptyObject, S extends TAnyObject = TAnyObject> {

  /**
   * Flag indicating whether current manager is still working or disposed.
   * Once manager is disposed, it cannot be reused or continue working.
   */
  public IS_DISPOSED: boolean = false;

  /**
   * Manager instance scope reference.
   * Used internally to emit signals/queries or subscribe to data.
   */
  public [SCOPE_SYMBOL]!: IScopeContext;
  /**
   * Signaling handler that operates scope signals and calls required method registered in metadata.
   */
  public [SIGNALING_HANDLER_SYMBOL]!: TSignalListener;

  /**
   * Manager instance signals metadata reference.
   * Used internally to emit signals with instance based description.
   */
  public [SIGNAL_METADATA_SYMBOL]!: TSignalSubscriptionMetadata;
  /**  /**
   * Manager instance signals metadata reference.
   * Used internally to emit signals with instance based description.
   */
  public [QUERY_METADATA_SYMBOL]!: TQuerySubscriptionMetadata;

  /**
   * Method allows to get related React.Context for manual renders or testing.
   * Lazy initialization, even for static resolving before anything from ContextManager is used.
   */
  public static get REACT_CONTEXT(): Context<any> {
    if (this === ContextManager) {
      throw new Error("Cannot reference to ContextManager statics directly. Only inherited classes allowed.");
    }

    return getReactContext(this as TConstructorKey);
  }

  /**
   * Manager instance context.
   * Generic field that will be synchronized with react providers when 'setContext' method is called.
   */
  public context: T = {} as T;

  /**
   * Generic context manager constructor.
   *
   * @param initialState - initial state received from dreamstate Provider component properties.
   *   Can be useful with external data provision or server side rendering.
   */
  public constructor(initialState?: S) {
    processComputed(this.context);
  }

  /**
   * Lifecycle method.
   * First provider was injected into react tree.
   * Same logic as 'componentWillMount' for class-based components.
   *
   * Useful for data initialization and subscriptions creation.
   */
  public onProvisionStarted(): void {}

  /**
   * Lifecycle method.
   * Last provider was removed from react tree.
   * Same logic as 'componentWillUnmount' for class-based components.
   *
   * Useful for data disposal when context is being ejected/when HMR happens.
   */
  public onProvisionEnded(): void {}

  /**
   * Forces update and render of subscribed components.
   * Just in case when you need forced update to keep everything in sync with your context.
   *
   * Note: creates new shallow copy of 'this.context' reference.
   */
  public forceUpdate(): void {
    /**
     * Always do shallow copy to point new ref object in current context.
     */
    this.context = processComputed(Object.assign({}, this.context));

    if (this[SCOPE_SYMBOL]) {
      this[SCOPE_SYMBOL].INTERNAL.notifyObservers(this);
    }
  }

  /**
   * Update current context from partially supplied state.
   * Updates react providers tree only if 'shouldObserversUpdate' check has passed and anything has changed in store.
   *
   * @param {Object|Function} next - part of context that should be updated or context transformer function.
   *   In case of functional callback, it will be executed immediately with 'currentContext' parameter.
   */
  public setContext(next: Partial<T> | TPartialTransformer<T>): void {
    const nextContext: T = Object.assign(
      {},
      this.context,
      /**
       * Handle context transformer functions.
       */
      isFunction(next) ? (next as TPartialTransformer<T>)(this.context) : next
    );

    /**
     * Always update context, even if it was created out of scope.
     * In case of existing scope just send additional notification.
     */
    this.context = processComputed(nextContext);

    /**
     * Compare current context with saved for observing one.
     */
    if (
      this[SCOPE_SYMBOL] &&
      shouldObserversUpdate(
        this[SCOPE_SYMBOL].INTERNAL.REGISTRY.CONTEXT_STATES_REGISTRY.get(
          this.constructor as TAnyContextManagerConstructor
        )!,
        nextContext
      )
    ) {
      this[SCOPE_SYMBOL].INTERNAL.notifyObservers(this);
    }
  }

  /**
   * Emit signal for other managers and subscribers in  current scope.
   *
   * @param {Object} baseSignal - signal base that contains basic descriptor of emitting signal.
   * @param {TSignalType} baseSignal.type - signal type.
   * @param {*=} baseSignal.data - optional signal data.
   * @returns {Promise} promise that will be resolved after signal listeners call.
   *   Note: async handlers will not be awaited.
   */
  public emitSignal<D = undefined>(baseSignal: IBaseSignal<D>): ISignalEvent<D> {
    if (this[SCOPE_SYMBOL]) {
      return this[SCOPE_SYMBOL].emitSignal(baseSignal, this.constructor as TAnyContextManagerConstructor);
    } else {
      throwOutOfScope();
    }
  }

  /**
   * Send context query to retrieve data from @OnQuery method with required types.
   */
  public queryDataAsync<D extends any, T extends TQueryType, Q extends IOptionalQueryRequest<D, T>>(
    queryRequest: Q
  ): Promise<TQueryResponse<any>> {
    if (this[SCOPE_SYMBOL]) {
      return this[SCOPE_SYMBOL].queryDataAsync(queryRequest);
    } else {
      throwOutOfScope();
    }
  }

  /**
   * Send sync context query to retrieve data from @OnQuery method with required types.
   */
  public queryDataSync<D extends any, T extends TQueryType, Q extends IOptionalQueryRequest<D, T>>(
    queryRequest: Q
  ): TQueryResponse<any> {
    if (this[SCOPE_SYMBOL]) {
      return this[SCOPE_SYMBOL].queryDataSync(queryRequest);
    } else {
      throwOutOfScope();
    }
  }

}
