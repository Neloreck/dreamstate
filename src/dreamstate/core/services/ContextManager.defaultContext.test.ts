import { mount } from "enzyme";
import { createElement, ReactElement } from "react";

import { useManager } from "@/dreamstate/core/consumption/useManager";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

function testManagers(first: TAnyContextManagerConstructor, second: TAnyContextManagerConstructor): void {
  function FirstConsumer(): ReactElement {
    const context = useManager(first);

    return createElement("div", {}, JSON.stringify(context));
  }

  function SecondConsumer(): ReactElement {
    const context = useManager(second);

    return createElement("div", {}, JSON.stringify(context));
  }

  expect(mount(createElement(FirstConsumer)).render()).toMatchSnapshot();
  expect(mount(createElement(SecondConsumer)).render()).toMatchSnapshot();
}

describe("ContextManager class default states", () => {
  it("should supply null values by default", () => {
    class First extends ContextManager {}

    class Second extends First {}

    testManagers(First, Second);
  });

  it("should supply default value if manager is not provided from static method, but handled in child", () => {
    class First extends ContextManager {}

    class Second extends First {

      public static getDefaultContext() {
        return { a: 1 };
      }

    }

    testManagers(First, Second);
  });

  it("should supply default value if manager is not provided from static method", () => {
    class First extends ContextManager {

      public static getDefaultContext() {
        return { a: 1 };
      }

    }

    class Second extends First {}

    testManagers(First, Second);
  });

  it("should supply default value from extended class method", () => {
    class First extends ContextManager<{ a: number }> {

      public static getDefaultContext() {
        return { a: 1 };
      }

    }

    class Second extends First {

      public static getDefaultContext() {
        return { a: 523 };
      }

    }

    testManagers(First, Second);
  });
});
