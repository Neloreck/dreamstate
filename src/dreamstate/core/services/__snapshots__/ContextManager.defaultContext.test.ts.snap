// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextManager class default states Should supply default value from extended class method 1`] = `
<div>
  {"a":1}
</div>
`;

exports[`ContextManager class default states Should supply default value from extended class method 2`] = `
<div>
  {"a":523}
</div>
`;

exports[`ContextManager class default states Should supply default value if manager is not provided from static method 1`] = `
<div>
  {"a":1}
</div>
`;

exports[`ContextManager class default states Should supply default value if manager is not provided from static method 2`] = `
<div>
  {"a":1}
</div>
`;

exports[`ContextManager class default states Should supply default value if manager is not provided from static method, but handled in child 1`] = `
<div>
  null
</div>
`;

exports[`ContextManager class default states Should supply default value if manager is not provided from static method, but handled in child 2`] = `
<div>
  {"a":1}
</div>
`;

exports[`ContextManager class default states Should supply null values by default 1`] = `
<div>
  null
</div>
`;

exports[`ContextManager class default states Should supply null values by default 2`] = `
<div>
  null
</div>
`;
